doc/wwwoffle-checkcert.man
.TH WWWOFFLE 8
.\" NAME should be all caps, SECTION should be 1-8, maybe w/ subsection
.\" other parms are allowed: see man(7), man(1)

.SH NAME
wwwoffle\-checkcert \- check wwwoffle root certificate

.SH SYNOPSIS
/usr/sbin/wwwoffle-checkcert

.SH DESCRIPTION
This utility checks whether the self-generated root certificate is expired,
or is about to expire. If the expiry will happen less than one week from
now, report this. If the certificate has already expired, also report this
and exit with return value 2.

.SH FILES
.br
.nf
\fI/etc/wwwoffle/certificates/root/root-key.pem\fR  key file
\fI/etc/wwwoffle/certificates/root/root-cert.pem\fR  certificate file
.fi

.SH AUTHOR
Written by Paul Slootman \fB<paul@debian.org>\fP




doc/NEWS
Version 2.9h of WWWOFFLE released : Mon Oct 3 13:00:00 2011
------------------------------------------------------------

Bug Fixes:
  Remove some gcc warnings.
  Delete keys & certs differently (fix crash with some gnutls versions). [*]
  Try decompressing buffered compressed data before asking for more.
  Handle buffered https data better. [*]

Improvements:
  Improve the SSL/https error messages (gnutls error conditions). [*]
  Use version 3 SSL certificates. [*]

*NOTE* *These only apply if WWWOFFLE is compiled with SSL/https support*


Version 2.9g of WWWOFFLE released : Sun Mar 27 11:00:00 2011
------------------------------------------------------------

Bug Fixes:
  Fix 2038 UNIX time_t problem with SSL certificates.
  Allow a margin on current time when warning about future timestamps.
  Check for libgnutls and libgcrypt separately at configure time.
  Better handling of the command line recursive options. 

New Features:
 A new option allows passing URLs from browser to server unmodified.


Version 2.9f of WWWOFFLE released : Sun Jan 31 19:00:00 2010
------------------------------------------------------------

Bug Fixes:
 Some compilation warnings caused by flex have been removed.
 When making a HEAD request use the cached version if possible.
 Don't use the libgnutls-config program in the configure script.
 Only remove "authorization" header if it is "basic" authentication.
 Send a 304 header when online if cached page won't be refreshed.
 Clarify SSL/HTTPS error messages and documentation.
 Clarify last online/offline messages from 'wwwoffle -status'.
 Perform MIME type tests case insensitively.

New Features:
 A new option to the wwwoffle program closes and re-opens the server log file.

Documentation:
 Several small fixes to documentation.

New SSL Features [*]:
 A faster but less cryptographically secure key generation option is provided.
 The expiration time of the SSL certificates can be increased from one year.

*NOTE* *These only apply if WWWOFFLE is compiled with SSL/https support*


Version 2.9e of WWWOFFLE released : Sun Jan 25 16:00:00 2009
------------------------------------------------------------

Bug Fixes:
 Handle deflate compression errors.  Don't allow wwwoffle program to make
 requests that won't be got.  Handle various encodings of URLs in HTML documents
 better.  Display a note on the monitor form if the page is already monitored.
 Stopped an error message when refreshing while online.

Documentation:
 Several small fixes to documentation.


Version 2.9d of WWWOFFLE released : Wed Jan 23 19:00:00 2008
------------------------------------------------------------

Bug Fixes:
 Don't output extra space before XHTML closing tag when modifying HTML.  Don't
 error in case of race condition creating directory.  Remove the generation and
 display of the XML format certificate.  Fix bug with decompressing zlib data.
 Ensure that creating files supplies a permission option.  Fetch image URLs
 stored in style attributes.

Documentation:
 Improve documentation of http-port and https-port configuration file options.

New Features:
 Add a new option to the CensorHeader section called referer-from.
 Add a new option to the FetchOptions section called iframes (default=no).


Version 2.9c of WWWOFFLE released : Sun Jul 15 14:30:00 2007
------------------------------------------------------------

Bug Fixes:
 Don't truncate the log file when opening it.  Fix some memory leaks and other
 lint problems.  Fix problem with info page for compressed cached files.  More
 fixes for 64-bit compilation.  Fix potential crash with case-insensitive
 wildcard matching.  Try to only get IPv4 socket info if no IPv6 addresses on
 system.  Better handling of https error if connection fails.  Add more
 information to fatal errors related to SSL certificates.


Version 2.9b of WWWOFFLE released : Fri Feb 16 09:30:00 2007
------------------------------------------------------------

Bug Fixes:
 Fix some string allocations.  Don't delete the lasttime cached spool file if
 page was fetched in this session.  Fix HTML parsing of '\' inside or outside
 quoted strings.  Make sure that files are truncated when opened.  Display
 non-empty value in monitor page hour-of-day setting for default value.  Some
 documentation updates.

New Features:
 Expired SSL certificates are deleted and replaced.


*NOTE* *This only applies if WWWOFFLE is compiled with SSL/https support*
       WWWOFFLE creates SSL certificates with a 1 year expiry period.  When they
       expire web browsers will start to give errors.  If the WWWOFFLE root
       Certificate Authority certificate expires then all sites will give SSL
       errors.  Deleting the existing WWWOFFLE CA certificate from the browser
       and installing the new one will fix this problem (the WWWOFFLE CA
       certificate can be loaded from https://localhost:8443/certificates/root).


Version 2.9a of WWWOFFLE released : Sun Aug 13 10:00:00 2006
------------------------------------------------------------

Bug Fixes:
 Fix crash with serving CA certificate.  On the info page show info for the alias
 if there is one.  Forced refresh headers take precedence over other options
 again.  Makefile config file installation error messages clarified.  Fix
 formatting in wwwoffle.conf manual page.  Handle compressed, chunked body of
 zero length.  Remove 'Keep-Alive' header.  Remove trailing spaces when parsing
 headers (fixes header censoring).  More fixes for 64-bit compilation.  Fix some
 allocated string lengths.  Keep '()' characters in attribute values when
 modifying HTML.

New Features:
 Added option 'disable-iframe' to ModifyHTML section to disable inline frames.

Programs:
 The wwwoffle-ls program lists only single URL unless a directory name is used.


*NOTE* The change to the wwwoffle-ls program means that the hyperestraier
       indexing script estwolefind may no longer work.  Edit estwolefind and

src/wwwoffle-checkcert.c
/***************************************
  Check certificate for expiry
  ******************/ /******************
  Written by Paul Slootman for the Debian wwwoffle package.
  Much copied from certificates.c, written and copyright by Andrew M. Bishop.

  This file Copyright 2010 Paul Slootman
  It may be distributed under the GNU Public License, version 2, or
  any higher version.  See section COPYING of the GNU Public license
  for conditions under which this file may be redistributed.
  ***************************************/


#include "autoconfig.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <unistd.h>

#include <sys/stat.h>
#include <fcntl.h>
#include <errno.h>

#include <gnutls/gnutls.h>
#include <gnutls/x509.h>
#include <gcrypt.h>

// #include "wwwoffle.h"
#include "errors.h"
#include "config.h"
#include "certificates.h"

/*+ Need this for Win32 to use binary mode +*/
#ifndef O_BINARY
#define O_BINARY 0
#endif

/*+ The number of bits for RSA private keys. +*/
#define RSA_BITS 512


/* Local functions */

static gnutls_x509_privkey_t    LoadPrivateKey(const char *filename);


/*+ The WWWOFFLE root certificate authority certificate. +*/
static gnutls_x509_crt_t root_crt;

/*+ The WWWOFFLE root certificate authority private key. +*/
static gnutls_x509_privkey_t root_privkey;

/* for errors.o: */
int gai_errno = 0;
int io_errno = 0;
int io_strerror = 0;

/*++++++++++++++++++++++++++++++++++++++
  Load in the WWWOFFLE root certificate authority certificate and private key.

  Check whether the certificate is usable.
  ++++++++++++++++++++++++++++++++++++++*/

int main(int argc, char*argv[])
{
 int err;
 struct stat buf;
 time_t activation,expiration,now;

 InitErrorHandler("wwwoffle-checkcert", 0, 1);

 if (chdir("/etc/wwwoffle") < 0)
     PrintMessage(Fatal, "Failed to chdir(/etc/wwwoffle)");

 /* Initialise the gnutls library */

 gnutls_global_init();

 /* Use faster but less secure key generation. */

 /* Read in the root private key in this directory. */

 if(access("certificates/root/root-key.pem",F_OK))
   {
    PrintMessage(Fatal,"The WWWOFFLE root CA private key file 'certificates/root/root-key.pem' does not exist.");
    exit(1);
   }

 root_privkey=LoadPrivateKey("certificates/root/root-key.pem");
 if(!root_privkey)
    PrintMessage(Fatal,"The WWWOFFLE root CA private key file 'certificates/root/root-key.pem' cannot be loaded; delete problem file and (re)start WWWOFFLE to recreate it.");

 /* Read in the root certificate in this directory. */

 if(access("certificates/root/root-cert.pem",F_OK))
   {
    PrintMessage(Fatal,"The WWWOFFLE root CA certificate file 'certificates/root/root-cert.pem' does not exist.");
   }

 root_crt=LoadCertificate("certificates/root/root-cert.pem");
 if(!root_crt)
    PrintMessage(Fatal,"The WWWOFFLE root CA certificate file 'certificates/root/root-cert.pem' cannot be loaded; delete problem file and (re)start WWWOFFLE to recreate it.");

 /* Check for expired root certificate and replace it */

 now=time(NULL);
 activation=gnutls_x509_crt_get_activation_time(root_crt);
 expiration=gnutls_x509_crt_get_expiration_time(root_crt);

 if(activation==-1 || expiration==-1 || activation>now || expiration<now)
   {
    PrintMessage(Fatal,"The WWWOFFLE root CA certificate has expired.");
   }

 if(expiration<now+7*86400)
   {
    if (expiration<now+2*86400)
      {
       PrintMessage(Warning,"The WWWOFFLE root CA certificate will expire in a day.");
      }
    else
      {
       PrintMessage(Warning,"The WWWOFFLE root CA certificate will expire in %d days.", expiration-now/86400);
      }
   }
 else
   {
    PrintMessage(Debug,"Expires: %d sec", expiration-now);
   }

 return(0);
}    


/*++++++++++++++++++++++++++++++++++++++
  Load in a single certificate from a file (the first if multiple).

  gnutls_x509_crt_t LoadCertificate Returns the loaded certificate.

  const char *filename The name of the file to load the certificate from.
  ++++++++++++++++++++++++++++++++++++++*/

gnutls_x509_crt_t LoadCertificate(const char *filename)
{
 gnutls_x509_crt_t *crt_list;
 int i=0;

 crt_list=LoadCertificates(filename);

 if(!crt_list || !crt_list[0])
    return(NULL);

 while(crt_list[++i])
    gnutls_x509_crt_deinit(crt_list[i]);

 return(crt_list[0]);
}


/*++++++++++++++++++++++++++++++++++++++
  Load in a certificate from a file.

  gnutls_x509_crt_t LoadCertificates Returns the loaded certificates (NULL terminated list).

  const char *filename The name of the file to load the certificates from.
  ++++++++++++++++++++++++++++++++++++++*/

gnutls_x509_crt_t *LoadCertificates(const char *filename)
{
 static gnutls_x509_crt_t crt_list[257];
 unsigned int n_crt=sizeof(crt_list)/sizeof(gnutls_x509_crt_t)-1;
 int fd,err;

 crt_list[0]=NULL;

 /* Load the certificates from the file. */

 fd=open(filename,O_RDONLY|O_BINARY);
 if(fd<0)
   {PrintMessage(Warning,"Could not open certificate file '%s' for reading [%!s].",filename);return(NULL);}
 else
   {
    struct stat buf;
    unsigned char *buffer;
    size_t buffer_size;
    gnutls_datum_t datum;

    if(fstat(fd,&buf))
      {PrintMessage(Warning,"Could not determine length of certificate file '%s' [%!s].",filename);return(NULL);}

    buffer=(unsigned char*)malloc(buf.st_size);
    buffer_size=buf.st_size;

    if(read(fd,buffer,buffer_size)!=buffer_size)
      {PrintMessage(Warning,"Could not read certificate file '%s' [%!s].",filename); free(buffer);return(NULL);}

    datum.data=buffer;
    datum.size=buffer_size;

    err=gnutls_x509_crt_list_import(crt_list,&n_crt,&datum,GNUTLS_X509_FMT_PEM,GNUTLS_X509_CRT_LIST_IMPORT_FAIL_IF_EXCEED);
    if(err<0)
      {PrintMessage(Warning,"Could not import certificates [%s].",gnutls_strerror(err)); free(buffer);return(NULL);}

    crt_list[err]=NULL;

    free(buffer);

    close(fd);
   }

 return(crt_list);
}


/*++++++++++++++++++++++++++++++++++++++
  Load in a private key from a file.

  gnutls_x509_privkey_t LoadPrivateKey Returns the loaded private key.

  const char *filename The name of the file to load the private key from.
  ++++++++++++++++++++++++++++++++++++++*/

static gnutls_x509_privkey_t LoadPrivateKey(const char *filename)
{
 gnutls_x509_privkey_t privkey;
 unsigned char buffer[2*RSA_BITS]; /* works for 256 bit keys or longer. */
 size_t buffer_size=sizeof(buffer);
 gnutls_datum_t datum;
 int fd,err;

 /* Load the private key from the file. */

 fd=open(filename,O_RDONLY|O_BINARY);
 if(fd<0)
   {PrintMessage(Warning,"Could not open private key file '%s' for reading [%!s].",filename);return(NULL);}
 else
   {
    buffer_size=read(fd,buffer,buffer_size);

    if(buffer_size==sizeof(buffer))
      {PrintMessage(Warning,"Could not read private key file '%s' [buffer not big enough].",filename);return(NULL);}

    datum.data=buffer;
    datum.size=buffer_size;

    err=gnutls_x509_privkey_init(&privkey);
    if(err<0)
      {PrintMessage(Warning,"Could not initialise private key [%s].",gnutls_strerror(err));return(NULL);}

    err=gnutls_x509_privkey_import(privkey,&datum,GNUTLS_X509_FMT_PEM);
    if(err<0)
      {PrintMessage(Warning,"Could not import private key [%s].",gnutls_strerror(err));return(NULL);}

    close(fd);
   }

 return(privkey);
}
